METRICS.md

```diff
@@ -0,0 +1,282 @@
+
+
+If we have a list of all changes for every deployment, it is simple to calculate the lead time for each change to deployment.
+
+```sql
+SELECT
+ d.deploy_id,
+ TIMESTAMP_TRUNC(d.time_created, DAY) as day,
+ ##### Time to Change
+ TIMESTAMP_DIFF(d.time_created, c.time_created, MINUTE) time_to_change_minutes
+FROM four_keys.deployments d, d.changes
+LEFT JOIN four_keys.changes c ON changes = c.change_id;
+```
+
+From this base, we want to extract the daily median lead time to change. 
+
+```sql
+SElECT
+day,
+PERCENTILE_CONT(
+  # Ignore automated changes
+  IF(time_to_change_minutes > 0,time_to_change_minutes, NULL), 
+  0.5) # Median
+  OVER (partition by day) median_time_to_change
+FROM
+(SELECT
+ d.deploy_id,
+ TIMESTAMP_TRUNC(d.time_created, DAY) as day,
+ # Time to Change
+ TIMESTAMP_DIFF(d.time_created, c.time_created, MINUTE) time_to_change_minutes
+ FROM four_keys.deployments d, d.changes
+ LEFT JOIN four_keys.changes c ON changes = c.change_id
+ )
+GROUP BY day, time_to_change_minutes;
+```
+
+It is up for debate whether or not we want to ignore the automated changes.  The rationale here is that when we merge a Pull Request it creates a Push event in the main branch.  This Push event is not its own distinct change, but rather a link in the workflow.  If we trigger a deployment off of this push event, this artificially skews the metrics and does not give us a clear picture of developer velocity. 
+
+### Calculating the bucket ###
+![Image of chart from the Four Keys dashboard, showing the median Lead Time to Change.](images/lead_time.png)
+
+To get the buckets, rather than aggregating daily, we look at the last 3 months and bucket the results according to the DORA research. 
+
+```sql
+SELECT 
+CASE WHEN median_time_to_change < 24 * 60 then "One day"
+     WHEN median_time_to_change < 168 * 60 then "One week"
+     WHEN median_time_to_change < 730 * 60 then "One month"
+     WHEN median_time_to_change < 730 * 6 * 60 then "Six months"
+     ELSE "One year"
+     END as lead_time_to_change,
+FROM
+  (SElECT
+    PERCENTILE_CONT(
+    # Ignore automated changes
+    IF(time_to_change_minutes > 0,time_to_change_minutes, NULL), 
+    0.5) # Median
+    OVER () median_time_to_change
+      FROM
+      (SELECT
+       d.deploy_id,
+       TIMESTAMP_TRUNC(d.time_created, DAY) as day,
+       # Time to Change
+       TIMESTAMP_DIFF(d.time_created, c.time_created, MINUTE) time_to_change_minutes
+       FROM four_keys.deployments d, d.changes
+       LEFT JOIN four_keys.changes c ON changes = c.change_id
+       # Limit to 3 months
+       WHERE d.time_created > TIMESTAMP(DATE_SUB(CURRENT_DATE(), INTERVAL 3 MONTH))
+       )
+      GROUP BY day, time_to_change_minutes
+      )
+LIMIT 1;
+```
+
+## Time to Restore Services ##
+
+**Definition**: For a failure, the median amount of time between the deployment which caused the failure and the restoration.  The restoration is measured by closing an associated bug / incident report. 
```

Comment:
I usually use the word "remediation" in this context, so to me, that would sound more natural. But if "restoration" is standard, then I will adapt to that.